Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,671 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambetr-Loir
+-/
+
+import Mathlib.Algebra.Exact
+import Mathlib.RingTheory.TensorProduct
+
+/-! # Right exactness properties of tensor product
+* `TensorProduct.rTensor_exact` says that one can tensor a short exact sequence on the right,
+  and `rTensor.equiv` gives a LinearEquiv.
+
+* `TensorProduct.lTensor_exact` says that one can tensor a short exact sequence on the left,
+  and `lTensor.equiv` gives a LinearEquiv.
+
+* For `N : Submodule R M`, `LinearMap.exact_subtype_mkQ N` says that
+  the inclusion of the submodule and the quotient map form an exact pair,
+  and `lTensor_mkQ` compute `ker (lTensor Q (N.mkQ))` and similarly for `rTensor_mkQ`
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map f g'`
+  in the presence of two short exact sequences.
+
+The proofs are those of [bourbaki1989] (chap. 2, §3, n°6)
+
+* Analogue for morphisms of algebras :
+  `Algebra.TensorProduct.ker_map` computes the kernel of `Algebra.TensorProduct.map f g`
+
+* The two more specific lemmas `Algebra.TensorProduct.lTensor_ker`
+  and `Algebra.TensorProduct.rTensor_ker` compute the kernels
+  of `Algebra.TensorProduct.map f id` and `Algebra.TensorProduct.map id g`
+
+## TODO
+
+* Treat the noncommutative case
+
+* Treat the case of modules over semirings
+  (For a possible definition of an exact sequence of commutative semigroups, see
+  [Grillet-1969b], Pierre-Antoine Grillet,
+  *The tensor product of commutative semigroups*,
+  Trans. Amer. Math. Soc. 138 (1969), 281-293, doi:10.1090/S0002-9947-1969-0237688-1 .)
+
+-/
+
+
+section Modules
+
+open TensorProduct LinearMap
+
+section Semiring
+
+variable {R : Type*} [CommSemiring R]
+variable {M N P P' Q: Type*}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommGroup P] [AddCommMonoid P'] [AddCommMonoid Q]
+  [Module R M] [Module R N] [Module R P] [Module R P'] [Module R Q]
+
+
+section Flip
+
+def TensorProduct.flip : M ⊗[R] N →ₗ[R] N ⊗[R] M :=
+  TensorProduct.lift (LinearMap.flip (TensorProduct.mk  R N M))
+
+lemma TensorProduct.flip_flip_eq :
+    (TensorProduct.flip : N ⊗[R] M →ₗ[R] M ⊗[R] N).comp TensorProduct.flip = LinearMap.id := by
+  unfold TensorProduct.flip
+  apply TensorProduct.ext'
+  intro x y
+  rfl
+
+def TensorProduct.flipEquiv : M ⊗[R] N ≃ₗ[R] N ⊗[R] M := {
```

Comment:
This already exists as `TensorProduct.comm`

Reply:
Cool! I'll use that then.